## 練習時間
假設我們資料中類別的數量並不均衡，在評估準確率時可能會有所偏頗，試著切分出 y_test 中，0 類別與 1 類別的數量是一樣的 (亦即 y_test 的類別是均衡的)

In [1]:
import numpy as np
X = np.arange(1000).reshape(200, 5)
y = np.zeros(200)
y[:40] = 1

In [2]:
from sklearn.model_selection import train_test_split, KFold
import numpy as np

In [3]:
X_train, X_test, y_train, y_test = train_test_split(X, y)


In [4]:
print('X: shape: ' + str(X.shape))
print(X)
print("")
print('y: shape: ' + str(y.shape))
print(y)

X: shape: (200, 5)
[[  0   1   2   3   4]
 [  5   6   7   8   9]
 [ 10  11  12  13  14]
 [ 15  16  17  18  19]
 [ 20  21  22  23  24]
 [ 25  26  27  28  29]
 [ 30  31  32  33  34]
 [ 35  36  37  38  39]
 [ 40  41  42  43  44]
 [ 45  46  47  48  49]
 [ 50  51  52  53  54]
 [ 55  56  57  58  59]
 [ 60  61  62  63  64]
 [ 65  66  67  68  69]
 [ 70  71  72  73  74]
 [ 75  76  77  78  79]
 [ 80  81  82  83  84]
 [ 85  86  87  88  89]
 [ 90  91  92  93  94]
 [ 95  96  97  98  99]
 [100 101 102 103 104]
 [105 106 107 108 109]
 [110 111 112 113 114]
 [115 116 117 118 119]
 [120 121 122 123 124]
 [125 126 127 128 129]
 [130 131 132 133 134]
 [135 136 137 138 139]
 [140 141 142 143 144]
 [145 146 147 148 149]
 [150 151 152 153 154]
 [155 156 157 158 159]
 [160 161 162 163 164]
 [165 166 167 168 169]
 [170 171 172 173 174]
 [175 176 177 178 179]
 [180 181 182 183 184]
 [185 186 187 188 189]
 [190 191 192 193 194]
 [195 196 197 198 199]
 [200 201 202 203 204]
 [205 206 207 208 209]
 [210 211 212 2

可以看見 y 類別中，有 160 個 類別 0，40 個 類別 1 ，請試著使用 train_test_split 函數，切分出 y_test 中能各有 10 筆類別 0 與 10 筆類別 1 。(HINT: 參考函數中的 test_size，可針對不同類別各自作切分後再合併)

In [5]:
kf = KFold(n_splits=10)
i = 0
#print(X[0:8])
#print(y[0:40])
for train_index, test_index in kf.split(X):
    i +=1 
    #print(train_index)
    #%print(test_index)
    x_train_1 ,x_test_1 , y_train_1   , y_test_1 = train_test_split(X[0:40] , y[0:40] )
    print("FOLD {}: ".format(i))
    print("X_test_1: ", x_test_1)
    print("X_train_1: ", x_train_1)
    print("Y_test_1: ", y_test_1)
    print("Y_train_1: ", y_train_1)
    print("-"*30)
    x_train_2 ,  x_test_2,y_train_2  , y_test_2 = train_test_split(X[40:] , y[40:] ,test_size=10)

    print("FOLD {}: ".format(i))
    print("X_test_2: ", x_test_2)
    print("X_train_2: ", x_train_2)
    print("Y_test_2: ", y_test_2)
    print("Y_train_2: ", y_train_2)
    print("-"*30)

FOLD 1: 
X_test_1:  [[105 106 107 108 109]
 [195 196 197 198 199]
 [120 121 122 123 124]
 [135 136 137 138 139]
 [180 181 182 183 184]
 [170 171 172 173 174]
 [ 35  36  37  38  39]
 [ 50  51  52  53  54]
 [185 186 187 188 189]
 [155 156 157 158 159]]
X_train_1:  [[145 146 147 148 149]
 [ 65  66  67  68  69]
 [175 176 177 178 179]
 [  0   1   2   3   4]
 [115 116 117 118 119]
 [125 126 127 128 129]
 [150 151 152 153 154]
 [ 95  96  97  98  99]
 [  5   6   7   8   9]
 [165 166 167 168 169]
 [ 70  71  72  73  74]
 [ 80  81  82  83  84]
 [110 111 112 113 114]
 [ 15  16  17  18  19]
 [ 45  46  47  48  49]
 [190 191 192 193 194]
 [140 141 142 143 144]
 [ 30  31  32  33  34]
 [ 20  21  22  23  24]
 [ 90  91  92  93  94]
 [ 10  11  12  13  14]
 [ 40  41  42  43  44]
 [ 25  26  27  28  29]
 [130 131 132 133 134]
 [ 55  56  57  58  59]
 [ 85  86  87  88  89]
 [100 101 102 103 104]
 [ 60  61  62  63  64]
 [ 75  76  77  78  79]
 [160 161 162 163 164]]
Y_test_1:  [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
Y_t


y_1_index, y_0_index = np.where(y==1)[0], np.where(y==0)[0] # 選出 y 等於 1 的 index 與 y 等於 0 的 index
X0_train, X0_test, y0_train, y0_test = train_test_split(X[0:40],y[0:40],random_state=42, test_size=10)
X1_train, X1_test, y1_train, y1_test = train_test_split(X[40:],y[40:], random_state=42, test_size=10)
print("-"*30)
print("X_test: ", X0_train)
print("X_train: ", X0_test)
print("Y_test: ", y0_train)
print("Y_train: ", y0_test)
print("-"*30)
print("X_test: ", X1_train)
print("X_train: ", X1_test)
print("Y_test: ", y1_train)
print("Y_train: ", y1_test)
print("-"*30)

In [8]:
y_test=np.concatenate([y_test_2, y_test_1])

In [9]:
print(y_test)

[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
